src/analysis/functions.rs

```diff
@@ -238,51 +260,201 @@ fn analyze_function(
 
-    for (pos, par) in parameters.c_parameters.iter().enumerate() {
-        assert!(
-            !par.instance_parameter || pos == 0,
-            "Wrong instance parameter in {}",
-            func.c_identifier.as_ref().unwrap()
-        );
-        if let Ok(s) = used_rust_type(env, par.typ, !par.direction.is_out()) {
-            used_types.push(s);
+    // Key: destroy callback index
+    // Value: associated user data index
+    let mut cross_user_data_check: HashMap<usize, usize> = HashMap::new();
+    let mut user_data_indexes: HashSet<usize> = HashSet::new();
+
+    if !expecting_data {
+        for (pos, par) in parameters.c_parameters.iter().enumerate() {
+            if async && par.c_type == "gpointer" {
+                continue;
+            }
+            assert!(
+                !par.instance_parameter || pos == 0,
+                "Wrong instance parameter in {}",
+                func.c_identifier.as_ref().unwrap()
+            );
+            if let Ok(s) = used_rust_type(env, par.typ, !par.direction.is_out()) {
+                used_types.push(s);
+            }
+            let (to_glib_extra, callback_info) = bounds.add_for_parameter(env, func, par, async, false);
+            if let Some(to_glib_extra) = to_glib_extra {
+                to_glib_extras.insert(pos, to_glib_extra);
+            }
+
+            analyze_async(
+                env,
+                func,
+                callback_info,
+                &mut commented,
+                &mut trampoline,
+                &mut async_future,
+            );
+            let type_error =
+                !(async && *env.library.type_(par.typ) == Type::Fundamental(library::Fundamental::Pointer)) &&
+                parameter_rust_type(env, par.typ, par.direction, Nullable(false), RefMode::None)
+                    .is_err();
+            if type_error {
+                commented = true;
+            }
         }
-        let (to_glib_extra, callback_info) = bounds.add_for_parameter(env, func, par, async);
-        if let Some(to_glib_extra) = to_glib_extra {
-            to_glib_extras.insert(pos, to_glib_extra);
+        if async {
+            if trampoline.is_none() {
+                commented = true;
+            }
         }
+    } else {
+        let mut to_replace = Vec::new();
+        let mut to_remove = Vec::new();
 
-        analyze_async(
-            env,
-            func,
-            callback_info,
-            &mut commented,
-            &mut trampoline,
-            &mut async_future,
-        );
-        let type_error =
-            !(async && *env.library.type_(par.typ) == Type::Fundamental(library::Fundamental::Pointer)) &&
-            parameter_rust_type(env, par.typ, par.direction, Nullable(false), RefMode::None)
-                .is_err();
-        if type_error {
+        {
+            // When closure data and destroy are specified in gir, they don't take into account the
+            // actual closure parameter.
```

Comment:
What does this mean? The counting of the indexes skips the closure parameter?

Are you sure this is not just because the counting *in general* does never include the instance parameter?

Reply:
It never counts the closure argument but it does count the instance parameter. Dark magic of gir and some pretty stupid way of working.

Reply:
That's indeed quite stupid